lib/spack/spack/cmd/ci.py

```diff
@@ -0,0 +1,492 @@
+
+    # Create a temporary working directory
+    with spack_ci.TemporaryDirectory() as temp_dir:
+        git = exe.which('git', required=True)
+
+        # Push a commit with the generated file to the downstream ci repo
+        saved_git_dir = os.path.join(temp_dir, 'original-git-dir')
+
+        shutil.move('.git', saved_git_dir)
+
+        git('init', '.')
+
+        git('config', 'user.email', 'robot@spack.io')
+        git('config', 'user.name', 'Spack Build Bot')
+
+        git('add', '.')
+
+        tty.msg('git commit')
+        commit_message = '{0} {1} ({2})'.format(
+            'Auto-generated commit testing', branch_name, commit_sha)
+
+        git('commit', '-m', '{0}'.format(commit_message))
+
+        tty.msg('git push')
+        git('remote', 'add', 'downstream', downstream_repo)
+        push_to_branch = 'master:multi-ci-{0}'.format(branch_name)
+        git('push', '--force', 'downstream', push_to_branch)
+
+        shutil.rmtree('.git')
+        shutil.move(saved_git_dir, '.git')
+        git('reset', '--hard', 'HEAD')
+
+
+def ci_rebuild(args):
+    """This command represents a gitlab-ci job, corresponding to a single
+       release spec.  As such it must first decide whether or not the spec it
+       has been assigned to build is up to date on the remote binary mirror.
+       If it is not (i.e. the full_hash of the spec as computed locally does
+       not match the one stored in the metadata on the mirror), this script
+       will build the package, create a binary cache for it, and then push all
+       related files to the remote binary mirror.  This script also
+       communicates with a remote CDash instance to share status on the package
+       build process.
+
+       The spec to be built by this job is represented by essentially two
+       pieces of information: 1) a root spec (possibly already concrete, but
+       maybe still needing to be concretized) and 2) a package name used to
+       index that root spec (once the root is, for certain, concrete)."""
+    env = ev.get_env(args, 'ci rebuild', required=True)
+
+    ci_artifact_dir = get_env_var('CI_PROJECT_DIR')
+    signing_key = get_env_var('SPACK_SIGNING_KEY')
+    spack_enable_cdash = get_env_var('SPACK_ENABLE_CDASH')
+    root_spec = get_env_var('SPACK_ROOT_SPEC')
+    remote_mirror_url = get_env_var('SPACK_MIRROR_URL')
+    enable_artifacts_mirror = get_env_var('SPACK_ENABLE_ARTIFACTS_MIRROR')
+    job_spec_pkg_name = get_env_var('SPACK_JOB_SPEC_PKG_NAME')
+    compiler_action = get_env_var('SPACK_COMPILER_ACTION')
+    cdash_base_url = get_env_var('SPACK_CDASH_BASE_URL')
+    cdash_project = get_env_var('SPACK_CDASH_PROJECT')
+    cdash_project_enc = get_env_var('SPACK_CDASH_PROJECT_ENC')
+    cdash_build_name = get_env_var('SPACK_CDASH_BUILD_NAME')
+    cdash_site = get_env_var('SPACK_CDASH_SITE')
+    related_builds = get_env_var('SPACK_RELATED_BUILDS')
+    job_spec_buildgroup = get_env_var('SPACK_JOB_SPEC_BUILDGROUP')
```

Comment:
Is the `DOWNSTREAM_CI_REPO` required here?

Reply:
It's required in order to trigger a subsequent pipeline with the generated workload.  We won't need this after gitlab implements child pipelines, supposedly coming out with version 12.6 later this month.

Because we push the generated `.gitlab-ci.yml` to a branch called `multi-ci-<originalBranchName>`, the downstream ci repo can actually be the same repo.

Reply:
Makes sense, I didn't see it in the list and thought it might need added.
Thanks for the details!